x-pack/elastic-agent/pkg/core/plugin/service/app.go

```diff
@@ -0,0 +1,330 @@
+			return err
+		}
+	}
+
+	defer func() {
+		if err != nil {
+			if a.srvState != nil {
+				a.srvState.Destroy()
+				a.srvState = nil
+			}
+		}
+	}()
+
+	if err := a.monitor.Prepare(a.name, a.pipelineID, a.uid, a.gid); err != nil {
+		return err
+	}
+
+	if a.limiter != nil {
+		a.limiter.Add()
+	}
+
+	// start the credentials listener for the service
+	if err := a.startCredsListener(); err != nil {
+		return err
+	}
+
+	// allow the service manager to ensure that the application is started, currently this does not start/stop
+	// the actual service in the system service manager
+
+	return nil
+}
+
+// Configure configures the application with the passed configuration.
+func (a *Application) Configure(_ context.Context, config map[string]interface{}) (err error) {
+	defer func() {
+		if err != nil {
+			// inject App metadata
+			err = errors.New(err, errors.M(errors.MetaKeyAppName, a.name), errors.M(errors.MetaKeyAppName, a.id))
+		}
+	}()
+
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	if a.srvState == nil {
+		return errors.New(ErrAppNotInstalled)
+	}
+
+	cfgStr, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.New(err, errors.TypeApplication)
+	}
+	err = a.srvState.UpdateConfig(string(cfgStr))
+	if err != nil {
+		return errors.New(err, errors.TypeApplication)
+	}
+	return nil
+}
+
+// Stop stops the current application.
+func (a *Application) Stop() {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	if a.srvState == nil {
+		return
+	}
+
+	if err := a.srvState.Stop(a.processConfig.StopTimeout); err != nil {
+		a.setState(state.Failed, errors.New(err, "Failed to stopped").Error())
+	} else {
+		a.setState(state.Stopped, "Stopped")
+	}
+	a.srvState = nil
+
+	a.cleanUp()
+	a.stopCredsListener()
+}
+
+// OnStatusChange is the handler called by the GRPC server code.
+//
+// It updates the status of the application and handles restarting the application is needed.
+func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.StateObserved_Status, msg string) {
+	a.appLock.Lock()
+
+	// If the application is stopped, do not update the state. Stopped is a final state
+	// and should not be overridden.
+	if a.state.Status == state.Stopped {
+		a.appLock.Unlock()
+		return
+	}
+
+	a.setStateFromProto(status, msg)
+	a.appLock.Unlock()
+}
+
+func (a *Application) setStateFromProto(pstatus proto.StateObserved_Status, msg string) {
+	var status state.Status
+	switch pstatus {
+	case proto.StateObserved_STARTING:
+		status = state.Starting
+	case proto.StateObserved_CONFIGURING:
+		status = state.Configuring
+	case proto.StateObserved_HEALTHY:
+		status = state.Running
+	case proto.StateObserved_DEGRADED:
+		status = state.Degraded
+	case proto.StateObserved_FAILED:
+		status = state.Failed
+	case proto.StateObserved_STOPPING:
+		status = state.Stopping
+	}
+	a.setState(status, msg)
+}
+
+func (a *Application) setState(status state.Status, msg string) {
+	if a.state.Status != status || a.state.Message != msg {
```

Comment:
do we need any locks in here?

Reply:
`setState` is a private function inside of the `Application` it is only ever called with the lock already held.